devicetree: improve output

mbolivar opened the pull request:
This makes the results a little easier to read, since node output appears in the same order as the top level comment summarizing the tree. It also makes gaps where nodes don't have matching compats more obvious.

mbolivar-nordic committed:
gen_defines: emit output for all nodes in dependency order

This makes the results easier to read, since node output appears in
the same order as the top level comment summarizing the tree.

Reorganize how flash partitions are emitted so they also appear in
dependency order, allowing us to extend the comment at the top of the
file to say that nodes are emitted in that order, along with other
useful information.

Print comments even for nodes that are disabled or have no binding,
to make it clearer that the script considered them and decided not to
output anything for them.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

ulfalizer reviewed (approved):
No strong opinions.

Maybe the comment at the top of the generated header could mention how the nodes are sorted.

pabigot reviewed (approved):
I agree this order is nicer to deal with.  It's not clear what the default order is, possibly depth-first or breadth-first.

I looked a little at possibly sorting the `node` property in this way, so the order would be consistent for all iterations, but that destabilizes the order of flash partitions.  So IMO it's best to leave this sort for this output.

mbolivar-nordic committed:
gen_defines: don't emit comments with "# " lines

If the line is empty, just emit a single "#\n" instead of "# \n".

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
> Maybe the comment at the top of the generated header could mention how the nodes are sorted.

I didn't want to do that because it wasn't totally true: the special casing around flash partitions is an exception, and I didn't want to take the time to fix it.

But you're right that that would be better, so it's fixed now. I'm also emitting comments for all nodes which should make it clearer why output for one might be missing (e.g. if there is no matching compatible or the node is disabled).

I tacked on what I consider a fix for DT_COMPAT_FIXED_PARTITIONS. I can split that into a separate PR if anyone wants.

mbolivar commented:
> I looked a little at possibly sorting the `node` property in this way, so the order would be consistent for all iterations, but that destabilizes the order of flash partitions. So IMO it's best to leave this sort for this output.

@pabigot note that the latest change has removed the property that flash partitions are sorted by start address, but I think it's better for consistency to keep it this way actually.

Can we add a tiebreaker to scc_order that considers the unit address when assigning ordinals to sibling nodes in the tree somehow? I haven't looked at the implementation.

mbolivar-nordic committed:
scripts: grutils: tweak sort keys to group siblings

Change the key function used to sort nodes so that unit addresses, if
present, break ties between sibling nodes. This orders siblings in
increasing order by unit-address in any gen_defines output that is
sorted by ordinal.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
> Can we add a tiebreaker to scc_order that considers the unit address when assigning ordinals to sibling nodes in the tree somehow? I haven't looked at the implementation.

Added a commit that does just that.

pabigot commented:
This still looks fine, modulo the need to rebase.  Thanks for making it more consistent.

mbolivar commented:
Rebased, thanks.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -67,11 +72,10 @@ def main():
         out(f"COMPAT_{str2ident(compat)}", 1)
 
-    # Derived from /chosen
+    # Definitions derived from /chosen nodes.
```
I skip the period when it's just one sentence, so there's a method to it at least.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -366,20 +386,15 @@ def write_addr_size(edt, prop_name, prefix):
 def write_flash(edt):
-    # Writes flash-related output
+    # Writes chosen and tree-wide flash-related output
+    global flash_area_num
```
I only put in `global` if it's written to, but fine if it was meant as a hint.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -792,7 +805,8 @@ def out_comment(s, blank_before=True):
         print("/* " + s + " */", file=header_file)
 
-    print("\n".join("# " + line for line in s.splitlines()), file=conf_file)
+    print("\n".join("#" + (' ' + line if line.strip() else '')
```
Maybe

```python
"# " + line if line.strip() else "#"
```

mbolivar commented:
@ulfalizer comments addressed, thanks!
